magenta/music/musicxml_parser.py

```diff
@@ -0,0 +1,516 @@
+    self.seconds = 0                # Duration in seconds
+    self.time_position = 0          # Onset time in seconds
+    self.voice = 1
+    self.is_rest = False
+    self.is_in_chord = False
+    self.is_grace_note = False
+    self.pitch = None               # Tuple (Pitch Name, MIDI number)
+    self.parse()
+
+  def parse(self):
+    global current_time_position
+    global previous_note
+
+    self.midi_channel = current_midi_channel
+    self.midi_program = current_midi_program
+    self.velocity = current_velocity
+
+    for child in self.xml_note:
+      if child.tag == "chord":
+        self.is_in_chord = True
+      elif child.tag == "duration":
+        self.duration = int(child.text)
+
+        self.midi_ticks = self.duration
+        self.midi_ticks *= (TICKS_PER_BEAT / current_divisions)
+
+        self.seconds = (self.midi_ticks / TICKS_PER_BEAT)
+        self.seconds *= current_seconds_per_beat
+
+        self.time_position = current_time_position
+
+        if self.is_in_chord:
+          # If this is a chord, subtract the duration from the
+          # previous note and make this note occur at the same time
+          previous_note.time_position -= self.seconds
+          self.time_position -= self.seconds
+        else:
+            # Only increment time positions once in chord
+            current_time_position += self.seconds
+
+      elif child.tag == "pitch":
+        self.parsePitch(child)
+      elif child.tag == "rest":
+        self.is_rest = True
+      elif child.tag == "voice":
+        self.voice = int(child.text)
+
+  def parsePitch(self, xml_pitch):
+    step = xml_pitch.find("step").text
+    alter = 0
+    if xml_pitch.find("alter") != None:
+      alter = xml_pitch.find("alter").text
+    octave = xml_pitch.find("octave").text
+
+    # Visual pitch representation
+    alter_string = ""
+    if alter == "-2":
+      alter_string = "bb"
+    elif alter == "-1":
+      alter_string = "b"
+    elif alter == "1":
+      alter_string = "#"
+    elif alter == "2":
+      alter_string = "x"
+
+    # N.B. - pitch_string does not account for transposition
+    pitch_string = step + alter_string + octave
+
+    # Compute MIDI pitch number (C4 = 60, C1 = 24, C0 = 12)
+    midi_pitch = self.pitchToMIDIPitch(step, alter, octave)
+    # Transpose MIDI pitch
+    midi_pitch = midi_pitch + current_transpose
+    self.pitch = (pitch_string, midi_pitch)
+
+  def pitchToMIDIPitch(self, step, alter, octave):
+    pitch_class = 0
+    if step == "C":
+      pitch_class = 0
+    elif step == "D":
+      pitch_class = 2
+    elif step == "E":
+      pitch_class = 4
+    elif step == "F":
+      pitch_class = 5
+    elif step == "G":
+      pitch_class = 7
+    elif step == "A":
+      pitch_class = 9
+    elif step == "B":
+      pitch_class = 11
+
+    pitch_class = (pitch_class + int(alter)) % 12
+    midi_pitch = (12 + pitch_class) + (int(octave) * 12)
+    return midi_pitch
+
+  def __str__(self):
```

Comment:
Is this method outputting a particular format, or just something to make debugging easier to read? Would probably be good to document in the docstring for this method.